x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import { EuiButtonEmpty, EuiCallOut, EuiFormRow, EuiSpacer, EuiText } from '@elastic/eui';
+import type { DataViewFieldBase } from '@kbn/es-query';
+import type { RequiredFieldInput } from '../../../../../common/api/detection_engine';
+import { UseArray, useFormData } from '../../../../shared_imports';
+import type { ArrayItem } from '../../../../shared_imports';
+import { RequiredFieldRow } from './required_fields_row';
+import * as ruleDetailsI18n from '../../../rule_management/components/rule_details/translations';
+import * as i18n from './translations';
+
+interface RequiredFieldsProps {
+  path: string;
+  indexPatternFields?: DataViewFieldBase[];
+  isIndexPatternLoading?: boolean;
+}
+
+export const RequiredFields = ({
+  path,
+  indexPatternFields = [],
+  isIndexPatternLoading = false,
+}: RequiredFieldsProps) => (
```

Comment:
Can this component and its children components be memoized to prevent unnecessary re-rendering that may impact page performance?

I've noticed, the more required fields added the more longer it takes for component to re-render

<img width="2546" alt="Screenshot 2024-05-08 at 15 51 08" src="https://github.com/elastic/kibana/assets/92328789/3836b09f-74ed-4479-af10-5a38c67fe999">

Just typing query in search bar, that seemingly does not have impact on required fields causes re-rendering that costs around 5ms, almost 20% of total form render.
With increasing number of fields it maybe even more impactful

Reply:
Hey Viltalii! I have investigated the issue thoroughly and ended up fixing it by:
 - wrapping the root `RequiredFields` component into a `React.memo`
 - passing the "watch" array to the `useFormData` hook. Please take a look at the comment and the implementation [here](https://github.com/nikitaindik/kibana/blob/editable-required-fields/x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/required_fields/required_fields.tsx#L80).

With these changes, the `RequiredFields` component no longer re-renders with every query change. Instead, it only re-renders when there are changes to either the "index patterns" or the "required fields".

Reply:
good job on tackling the issue

<img width="2556" alt="Screenshot 2024-05-15 at 09 46 31" src="https://github.com/elastic/kibana/assets/92328789/1922c489-2b81-4f2d-a9a2-4bb53a8f6cc1">